presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSchemaEvolutionUtils.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.hudi;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.hive.HiveType;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.serde2.typeinfo.DecimalTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.ListTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.MapTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.StructTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.model.HoodieTableType;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.TableSchemaResolver;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.table.timeline.HoodieTimeline;
+import org.apache.hudi.common.util.InternalSchemaCache;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.internal.schema.InternalSchema;
+import org.apache.hudi.internal.schema.Type;
+import org.apache.hudi.internal.schema.Types;
+import org.apache.hudi.internal.schema.action.InternalSchemaMerger;
+import org.apache.hudi.internal.schema.utils.InternalSchemaUtils;
+import org.apache.hudi.internal.schema.utils.SerDeHelper;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.hive.HiveType.toHiveType;
+
+public final class HudiSchemaEvolutionUtils
+{
+    private static final Logger log = Logger.get(HudiSchemaEvolutionUtils.class);
+
+    private HudiSchemaEvolutionUtils()
+    {
+    }
+
+    public static SchemaEvolutionContext createSchemaEvolutionContext(HoodieTableMetaClient metaClient)
+    {
+        // no need to do schema evolution for mor table, since hudi kernel will do it.
+        if (metaClient.getTableType() == HoodieTableType.MERGE_ON_READ || metaClient == null) {
+            return new SchemaEvolutionContext("", "");
+        }
+
+        try {
+            TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);
+            String internalSchema = schemaUtil.getTableInternalSchemaFromCommitMetadata().map(SerDeHelper::toJson).orElse("");
+            HoodieTimeline hoodieTimeline = metaClient.getCommitsAndCompactionTimeline().filterCompletedInstants();
+            String validCommits = hoodieTimeline.getInstants().map(HoodieInstant::getFileName).collect(Collectors.joining(","));
```

Comment:
> I guess it will be good to have some test cases covering the scenarios of different types of schema evolutions. That should clear most of my doubts as well.

agree , as https://github.com/apache/hudi/pull/6989  has merged in hudi kernel, we should add test cases to convering schema evolution.